custom_components/solplanet/services.py

```diff
@@ -0,0 +1,162 @@
+"""Services for Solplanet integration."""
+from __future__ import annotations
+
+import logging
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+import voluptuous as vol
+
+from .client import ScheduleSlot, BatterySchedule
+from .const import DOMAIN, BATTERY_IDENTIFIER
+
+DAYS = ["Mon", "Tus", "Wen", "Thu", "Fri", "Sat", "Sun"]
+
+_LOGGER = logging.getLogger(__name__)
+
+async def get_isn_from_target(hass: HomeAssistant, target: dict) -> list[str]:
+    """Get ISNs from target entity_ids or device_ids."""
+    isns = set()
+    
+    # Handle entity_ids
+    if "entity_id" in target:
+        entity_reg = er.async_get(hass)
+        entity_ids = target["entity_id"] if isinstance(target["entity_id"], list) else [target["entity_id"]]
+            
+        for entity_id in entity_ids:
+            if entry := entity_reg.async_get(entity_id):
+                parts = entry.unique_id.split('_')
+                if len(parts) > 2:
+                    isns.add(parts[2])
+                
+    # Handle device_ids
+    if "device_id" in target:
+        device_reg = dr.async_get(hass)
+        device_ids = target["device_id"] if isinstance(target["device_id"], list) else [target["device_id"]]
+            
+        for device_id in device_ids:
+            if device := device_reg.async_get(device_id):
+                for identifier in device.identifiers:
+                    if identifier[0] == DOMAIN:
+                        isn = identifier[1].replace("battery_", "")
+                        isns.add(isn)
+                        break
+
+    return list(isns)
+
+async def async_setup_services(hass: HomeAssistant) -> None:
+    """Set up services for Solplanet integration."""
+
+    async def set_schedule_slot(call: ServiceCall) -> None:
+        """Handle set_schedule_slot service."""
+        target = call.target if hasattr(call, 'target') else {}
+        if 'entity_id' in call.data:
+            target['entity_id'] = call.data['entity_id']
+        if 'device_id' in call.data:
+            target['device_id'] = call.data['device_id']
+
+        isns = await get_isn_from_target(hass, target)
+        if not isns:
+            raise vol.Invalid("No valid entities or devices found")
+            
+        processed = False
+        for isn in isns:
+            for entry_id, data in hass.data[DOMAIN].items():
```

Comment:
`entry_id` is not used

```suggestion
            for data in hass.data[DOMAIN].values():
```